File: lab4/lang-tests/test-5/tail_recursion_optimized.scala
object Assignment {

    @tailrec def factorialAcc(acc: Int, n: Int): Int = {
        if (n <= 1) 1/0
        else factorialAcc(n * acc, n - 1)
    }

    def main (args: Array[String]) {
        factorialAcc(1, 10);
    }
}

File: lab4/lang-tests/test-5/tail_recursion_optimized.java
class TailRec {
    public static int t_rec(int n, int acc) {
        if (n == 0) {
            for (StackTraceElement ste : Thread.currentThread().getStackTrace()) 
            {
                System.out.println(ste);
            }
            return 1+acc;
        }
        else {
            return t_rec(n-1, acc+1);
        }
    }
    public static void main(String[] args) {
        int res = t_rec(10, 0);
        System.out.println("t_rec(10) =>"+res);
    }
}